lnst/Recipes/ENRT/ConfigMixins/FirewallMixin.py

```diff
@@ -0,0 +1,197 @@
+    different ruleset sizes.
+    """
+
+    @property
+    def bulk_firewall_rules_generator(self):
+        """
+        A generator yielding { host: ruleset, ... } to apply for a test run.
+        Each yield will turn into a new sub configuration and thus be tested
+        separately.
+        """
+        raise NotImplementedError("Method bulk_firewall_rules_generator must be defined by a child class.")
+
+    def generate_sub_configurations(self, config):
+        for ruleset in self.bulk_firewall_rules_generator:
+            new_config = copy.copy(config)
+            new_config.firewall_rulesets = ruleset
+            yield new_config
+
+class NftablesMixin(FirewallMixin):
+    """
+    An nftables backend for FirewallMixin.
+    """
+
+    def _apply_ruleset(self, host, path):
+        host.run('nft flush ruleset')
+        host.run('nft -f "%s"' % path)
+
+    def _dump_ruleset(self, host, config):
+        return host.run('nft list ruleset').stdout
+
+class Ip46tablesMixin(FirewallMixin):
+    """
+    A common base class for all the iptables-like FirewallMixin backends.
+    Do not inherit directly, use one of the *tablesMixin classes instead.
+    """
+
+    @property
+    def iptables_command(self):
+        raise NotImplementedError("Property iptables_command must be defined by a child class.")
+
+    def _apply_ruleset(self, host, path):
+        host.run('%s-restore "%s"' % (self.iptables_command, path))
+
+    def _extract_tables_n_chains(self, ruleset):
+        out = {}
+        curtable = None
+        for line in ruleset.split('\n'):
+            if line[0] == '*':
+                out[line] = []
+            elif line[0] == ':':
+                words = line.split(' ')
+                if not words[1] == '-': # ignore user-defined chains
+                    out[curtable].append(words[0][1:])
+        return out
+
+    def _append_missing_parts(self, dst_ruleset, src_ruleset):
+        for table, chains in self._extract_tables_n_chains(src_ruleset).items():
+            tline = '*' + table + '\n'
+            if dst_ruleset.find(tline) >= 0:
+                continue
+            dst_ruleset += tline
+            for chain in chains:
+                dst_ruleset += ':' + chain
+                dst_ruleset += ' ACCEPT [0:0]\n'
+            dst_ruleset += 'COMMIT\n'
+        return dst_ruleset
+
+    def _dump_ruleset(self, host, config):
+        out = host.run('%s-save' % self.iptables_command).stdout
+        return self._append_missing_parts(out, config.firewall_rulesets[host])
+
+    def remove_sub_configuration(self, config):
+        """
+        override parent's implementation:
+        If the stored ruleset does not contain all tables of the current one,
+        rules remain in place.
+        """
+        for host, ruleset in config.firewall_rulesets.items():
+            stored = self.stored_rulesets[host]
+            self.stored_rulesets[host] = self.append_missing_parts(stored, ruleset)
```

Comment:
You're missing underscore in `self.append_missing_parts`

Reply:
Indeed, thanks!